internal/lifecycle/lifecycle.go

```diff
@@ -72,26 +72,31 @@ func (l *Lifecycle) Append(hook Hook) {
 func (l *Lifecycle) Start(ctx context.Context) error {
 	l.startRecords = make(HookRecords, 0, len(l.hooks))
 	for _, hook := range l.hooks {
-		if hook.OnStart != nil {
-			l.logger.LogEvent(&fxevent.LifecycleHookStart{CallerName: hook.callerFrame.Function})
-
-			l.mu.Lock()
-			l.runningHook = hook
-			l.mu.Unlock()
-
-			begin := time.Now()
-			if err := hook.OnStart(ctx); err != nil {
-				return err
+		select {
+		case <-ctx.Done():
```

Comment:
Since you have a return in your `case <-ctx.Done():`, you could git rid of `default:` and pull the rest out of the select

Reply:
So in that case, `select` will block till `ctx.Done()`, which is not what we want. We need these under `default:` so that when `ctx.Done()` hasn't yet happened, we keep executing the hooks. The select on `ctx.Done()` here is specifically for the case when `ctx` is cancelled/timed out while we're in the middle of executing the hooks, we can bail before we finish executing all the hooks instead of waiting for all of them.

Reply:
Makes sense, thanks for clarifying!